### PR TITLE
Migrate JamesServerTests to junit 5

### PR DESCRIPTION
**Pull request by trantienduchn**
server for #1749 (need JamesServerTest under Junit 5 extension -> easier to apply rabbitmq extension onto junit 5 JamesServerTest )

### CODE REVIEW

**Comment by trantienduchn** [comment]
```
[9720673380d03fc7baaf9d7b3233b35fd9659572] [ERROR] Failures: 
[9720673380d03fc7baaf9d7b3233b35fd9659572] [ERROR]   DistantMailboxPathRegisterTest.unregisterMixedWithRegisterShouldWorkInAConcurrentEnvironment:328 
[9720673380d03fc7baaf9d7b3233b35fd9659572] Expecting:
[9720673380d03fc7baaf9d7b3233b35fd9659572]  <{namespace:user:name=200L}>
[9720673380d03fc7baaf9d7b3233b35fd9659572] to contain:
[9720673380d03fc7baaf9d7b3233b35fd9659572]  <[MapEntry[key=namespace:user:name, value=100L]]>
[9720673380d03fc7baaf9d7b3233b35fd9659572] but could not find:
[9720673380d03fc7baaf9d7b3233b35fd9659572]  <[MapEntry[key=namespace:user:name, value=100L]]>
```
test this please

**Comment by trantienduchn** [comment]
```
[9720673380d03fc7baaf9d7b3233b35fd9659572] [ERROR] Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 53.459 s <<< FAILURE! - in org.apache.james.CassandraLdapJamesServerTest
[9720673380d03fc7baaf9d7b3233b35fd9659572] [ERROR] connectJMAPServerShouldRespondBadRequest  Time elapsed: 39.191 s  <<< ERROR!
[9720673380d03fc7baaf9d7b3233b35fd9659572] java.net.ConnectException: Connection refused (Connection refused)
```

test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
🤔 
```
[33eae78290a9deb97ffd772a002b30fde8f54126] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project james-server-cassandra-guice: There was a timeout or other error in the fork -> [Help 1]
```

**Comment by trantienduchn** [comment]
```
[66460c912efe2781b574dd19c61a80ebb7758b41] [ERROR]   ESReporterTest.timeMetricsShouldBeReportedWhenImapCommandsReceived:111 » ConditionTimeout
[66460c912efe2781b574dd19c61a80ebb7758b41] [ERROR]   ESReporterTest.timeMetricsShouldBeReportedWhenJmapRequestsReceived:133 » ConditionTimeout
```

FYI: ESReporterTest runs fine in single test. But not ok in `mvn test` damn

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
while executing `ESReporterTest`, ci printed many logs like this:

```
[8be841a14ce2ef92a5e1ccfb73f736d41ef9dc5c] 12:03:08.346 [DEBUG] c.d.d.c.Q.SLOW - [cluster53] [/192.168.0.1:55077] Query too slow, took 18 ms (99.0 percentile = 2 ms): SELECT * FROM firstUnseen LIMIT 1;
[8be841a14ce2ef92a5e1ccfb73f736d41ef9dc5c] 12:03:08.346 [DEBUG] c.d.d.c.Q.SLOW - [cluster53] [/192.168.0.1:55077] Query too slow, took 18 ms (99.0 percentile = 2 ms): SELECT * FROM domains LIMIT 1;
[8be841a14ce2ef92a5e1ccfb73f736d41ef9dc5c] 12:03:08.346 [DEBUG] c.d.d.c.Q.SLOW - [cluster53] [/192.168.0.1:55077] Query too slow, took 18 ms (99.0 percentile = 2 ms): SELECT * FROM annotation LIMIT 1;
[8be841a14ce2ef92a5e1ccfb73f736d41ef9dc5c] 12:03:08.360 [DEBUG] c.d.d.c.Q.SLOW - [cluster53] [/192.168.0.1:55077] Query too slow, took 13 ms (99.0 percentile = 2 ms): TRUNCATE domains;
[8be841a14ce2ef92a5e1ccfb73f736d41ef9dc5c] 12:03:08.364 [DEBUG] c.d.d.c.Q.SLOW - [cluster53] [/192.168.0.1:55077] Query too slow, took 18 ms (99.0 percentile = 2 ms): TRUNCATE 
```

I don't have any idea. The logs doesn't appear on my local machine, even in `mvn test`

**Comment by trantienduchn** [comment]
```
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR] Errors: 
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldCreateMessageInOutboxWhenSendingMessage:2264 NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldDeliverMessageToRecipient:2697 » NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldNotAllowCopyWhenOverQuota:1398 NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldNotAllowDraftCreationWhenOverQuota:1283 NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldRenderCorrectlyInIMAPMultivaluedUserAddedHeaders:4965 NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldSendAReadableHtmlMessage:2968 » NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldSendAReadableTextPlusHtmlMessage:3112 » NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldStripBccFromDeliveredEmail:2787 » NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldTriggerMaxQuotaReachedWhenTryingToSendMessageAndQuotaReached:2732 NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesShouldVerifyHeaderOfMessageInInbox:4369 » NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.setMessagesWhenSavingToDraftsShouldNotSendMessage:3007 » NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SetMessagesMethodTest.textBodyOfMessageWithTextCalendarShouldBeConvertedToAttachment:5357 NullPointer
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:230 » ConditionTimeout
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:136 » ConditionTimeout
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [INFO] 
[808faac8dbf7af4acd53c6f674bf9a86e4df2781] [ERROR] Tests run: 652, Failures: 0, Errors: 14, Skipped: 1
```

**Comment by trantienduchn** [comment]
```
[838b185127c583b78b7f23de1867d95b55404f19] [ERROR] Errors: 
[838b185127c583b78b7f23de1867d95b55404f19] [ERROR]   JmapJamesServerContract.connectJMAPServerShouldRespondBadRequest » NoClassDefFound
[838b185127c583b78b7f23de1867d95b55404f19] [ERROR]   CassandraLdapJamesServerTest.mailsShouldBeWellReceivedBeforeFirstUserConnectionWithLdap » NoClassDefFound
[838b185127c583b78b7f23de1867d95b55404f19] [ERROR]   CassandraLdapJamesServerTest.userFromLdapShouldLoginViaImapProtocol » NoClassDefFound
[838b185127c583b78b7f23de1867d95b55404f19] [ERROR]   JmapJamesServerContract.connectJMAPServerShouldRespondBadRequest » NoClassDefFound
```

**Comment by trantienduchn** [comment]
```
[f1e724c8fc0cdc271a7cc9bfe441d044fcf8dec9] [ERROR] Errors: 
[f1e724c8fc0cdc271a7cc9bfe441d044fcf8dec9] [ERROR]   RabbitMQClusterTest$ClusterNodesFailure.setup:165 » IO
```
test this please

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapJamesServerTest.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
     @ClassRule
     public static DockerCassandraRule cassandra = new DockerCassandraRule();
-    
+
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapJamesServerTest.java]: please avoid this kind of noise

File: server/container/guice/memory-guice/pom.xml [hunk 2]
```diff
@@ -153,8 +153,13 @@
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
+            <groupId>org.junit.vintage</groupId>
+            <artifactId>junit-vintage-engine</artifactId>
```
Comment (mbaechler) [inline server/container/guice/memory-guice/pom.xml]: vintage should come after jupiter

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java [hunk 3]
```diff
@@ -28,82 +28,72 @@
...
+    public void tearDown() throws IOException {
+        this.messageSender.close();
+        this.imapMessageReader.close();
+        this.imapClient.close();
     }
 
     @Test
-    public void userFromLdapShouldLoginViaImapProtocol() throws Exception {
-        imapClient.connect(JAMES_SERVER_HOST, server.getProbe(ImapGuiceProbe.class).getImapPort());
+    void userFromLdapShouldLoginViaImapProtocol() throws Exception {
```
Comment (mbaechler) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: what about injection jamesServer as a method argument ?

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestExtension.java [hunk 4]
```diff
@@ -22,37 +22,76 @@
...
+import org.junit.jupiter.api.extension.ExtensionContext;
 import org.junit.rules.TemporaryFolder;
-import org.junit.rules.TestRule;
-import org.junit.runner.Description;
-import org.junit.runners.model.Statement;
 
 import com.google.inject.Module;
 
-public class CassandraLdapJmapTestRule implements TestRule {
+class CassandraLdapJmapTestExtension implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback {
```
Comment (mbaechler) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestExtension.java]: it should compose on top of CassandraJmapTestExtension

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java [hunk 5]
```diff
@@ -39,47 +38,117 @@
...
+            this.additionalModules = ImmutableList.copyOf(additionalModules);
+            return this;
+        }
+
+        public Builder extensions(ImmutableList<GuiceModuleTestExtension> extensions) {
+            this.extensions = extensions;
+            return this;
+        }
+
+        public Builder ignoreEach() {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java]: what does it mean ?

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerElasticSearchTestExtension.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        elasticSearchContainer.start();
+    }
+
+    @Override
+    public void afterAll(ExtensionContext extensionContext) throws Exception {
+        elasticSearchContainer.stop();
+    }
+
+    @Override
+    public void afterEach(ExtensionContext extensionContext) throws Exception {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerElasticSearchTestExtension.java]: you probably don't use the good composition model if you need these empty methods

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java (hunk 3 above)
Comment (trantienduchn) [reply server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: please checkout #1759

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java (hunk 5 above)
Comment (trantienduchn) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java]: please checkout #1759

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerElasticSearchTestExtension.java (hunk 6 above)
Comment (trantienduchn) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/DockerElasticSearchTestExtension.java]: idem

**Comment by trantienduchn** [comment]
separate in smaller PR in #1759

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java (hunk 5 above)
Comment (mbaechler) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java]: if you need to explain to me the intent of a piece of code, either I missed something obvious or the intent is not clear enough. In the second case, you then need to "express your intent better"(™) (cc @ddolcimascolo )

Comment (trantienduchn) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraJmapTestExtension.java]: sorry I will remove these lines in #1759, and added again when the code changes need it.

**Pull request closed without merging**